lib/backend/watchersyncer/watchersyncer.go

```diff
@@ -0,0 +1,200 @@
+}
+
+func (rs *watcherSyncer) Start() {
+	log.Info("Start called")
+	go rs.run()
+}
+
+// Send a status update and store the status.
+func (rs *watcherSyncer) sendStatusUpdate(status api.SyncStatus) {
+	log.WithField("Status", status).Info("Sending status update")
+	rs.callbacks.OnStatusUpdated(status)
+	rs.status = status
+}
+
+// run implements the main syncer loop that loops forever receiving watch events and translating
+// to syncer updates.
+func (rs *watcherSyncer) run() {
+	log.Info("Running - sending status update")
+	rs.sendStatusUpdate(api.WaitForDatastore)
+	log.Info("Starting watcher caches")
+	for _, wc := range rs.watcherCaches {
+		go wc.run()
+	}
+
+	log.Info("Starting main watch processing loop")
+	var updates []api.Update
+	for {
+		// Block until there is data.
+		result := <-rs.results
+
+		// Process the data - this will append the data in subsequent calls, and action
+		// it if we hit a non-update event.
+		updates := rs.processResult(updates, result)
+
+		// Append results into the one update until we either flush the channel or we
+		// hit our fixed limit per update.
+	consolidatationloop:
+		for ii := 0; ii < maxUpdatesToConsolidate; ii++ {
+			select {
+			case next := <-rs.results:
+				updates = rs.processResult(updates, next)
+			default:
+				break consolidatationloop
+			}
+		}
+
+		// Perform final processing (pass in a nil result) before we loop and hit the blocking
+		// call again.
+		updates = rs.processResult(updates, nil)
+	}
+
+}
+
+// Process a result from the result channel.  We don't immediately action updates, but
+// instead start grouping them together so that we can send a larger single update to
+// Felix.
+func (rs *watcherSyncer) processResult(updates []api.Update, result interface{}) []api.Update {
+
+	// Final call of the interation, send any updates if we've collated some.
+	if result == nil {
+		if len(updates) > 0 {
+			rs.callbacks.OnUpdates(updates)
+		}
+		return nil
+	}
+
+	// Switch on the result type.
+	switch r := result.(type) {
+	case []api.Update:
+		// This is an update.  If we don't have previous updates then also check to see
+		// if we need to shift the status into Resync.
+		// We append these updates to the previous if there were any.
+		if len(updates) == 0 {
+			if rs.status == api.WaitForDatastore {
+				rs.sendStatusUpdate(api.ResyncInProgress)
+			}
+			updates = r
+		} else {
+			updates = append(updates, r...)
+		}
+
+	case error:
+		// Received an error.  Firstly, send any updates that we have grouped.
+		if len(updates) > 0 {
+			rs.callbacks.OnUpdates(updates)
+			updates = nil
+		}
+
+		// If this is a parsing error, and if the callbacks support
+		// it, then send the error update.
+		log.WithError(r).Info("Error received in syncer")
+		if ec, ok := rs.callbacks.(api.SyncerParseFailCallbacks); ok {
+			if pe, ok := r.(cerrors.ErrorParsingDatastoreEntry); ok {
+				ec.ParseFailed(pe.RawKey, pe.RawValue)
+			}
+		}
+
+	case api.SyncStatus:
```

Comment:
Just so I understand, this code is built on the assumption that the individual watcherCaches can send no other types besides `api.InSync`

If so, it would be good to have an assert or something here - I can imagine one day someone will try to add a new call which throws off the value here.

Reply:
Rather than assert, I log the status we receive and no op or anything other than the InSync